### PR TITLE
Removing all FakeApp usages

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
Removing all FakeApp usages

Related PR: mattermost/enterprise#602
Jira Ticket: https://mattermost.atlassian.net/browse/MM-25908

### CODE REVIEW

**Comment by agnivade** [comment]
Thank you for this !! :scream: :tada:

File: app/app.go [hunk 1]
```diff
@@ -63,6 +62,66 @@ func New(options ...AppOption) *App {
 	return app
 }
 
+func (a *App) InitServer() {
+	a.srv.AppInitialized.Do(func() {
```
Comment (jespino) [inline app/app.go]: This is one of the key changes. I running this the first time I instantiate the App, so we initialize the server part that is strictly dependent on the App itself.

**Comment by agnivade** [comment]
This is big one and obviously has far reaching impacts. I will take some time to review this.

**Comment by jespino** [comment]
@agnivade sure, take your time, no rush :)

File: app/server.go [hunk 2]
```diff
@@ -48,6 +48,7 @@ var MaxNotificationsPerChannelDefault int64 = 1000000
 type Server struct {
 	Store           store.Store
 	WebSocketRouter *WebSocketRouter
+	AppInitialized  sync.Once
```
Comment (agnivade) [inline app/server.go:51]: Does this need to be exported ?

Idiomatic Go also suggests that it should be named - "appInitializedOnce".

File: app/app.go [hunk 3]
```diff
@@ -63,6 +63,66 @@ func New(options ...AppOption) *App {
...
+		a.initJobs()
+		a.srv.RunJobs()
+	})
+	a.accountMigration = a.srv.AccountMigration
+	a.ldap = a.srv.Ldap
+	a.notification = a.srv.Notification
+	a.saml = a.srv.Saml
+}
+
 // DO NOT CALL THIS.
```
Comment (agnivade) [inline app/app.go:126]: Unrelated, but can we also file a ticket to fix this sometime ? Would be good to get it done.

File: app/enterprise_test.go [hunk 4]
```diff
@@ -116,6 +116,7 @@ func TestSAMLSettings(t *testing.T) {
 			}
 
 			th.Server.initEnterprise()
+			th.App.initEnterprise()
```
Comment (agnivade) [inline app/enterprise_test.go:119]: Hmm .. wondering if we can do the server init inside the app init ? The `srv` field is inside App anyways.

File: app/cluster_discovery.go [hunk 5]
```diff
@@ -16,38 +16,42 @@ const (
 
 type ClusterDiscoveryService struct {
 	model.ClusterDiscovery
-	app  *App
-	stop chan bool
+	server *Server
```
Comment (agnivade) [inline app/cluster_discovery.go:19]: Would suggest it to be `srv` to maintain consistency with the naming.

**Comment by agnivade** [comment]
Btw, this is going to create a nasty merge conflict with my websocket refactor :laughing:

File: app/cluster_discovery.go (hunk 5 above)
Comment (jespino) [reply app/cluster_discovery.go:19]: :+1:

File: app/enterprise_test.go (hunk 4 above)
Comment (jespino) [reply app/enterprise_test.go:121]: For testing, makes sense.

File: app/app.go (hunk 3 above)
Comment (jespino) [reply app/app.go:126]: sure we should

File: app/options.go [hunk 6]
```diff
@@ -105,11 +105,8 @@ func ServerConnector(s *Server) AppOption {
 		a.compliance = s.Compliance
```
[the remark this answers is not in the document]
Comment (jespino) [reply app/options.go:105]: We don't necesarrilly want to get rid of this, the App is going to wrap the server anyway, so connect the server to the new app instances, is correct, maybe not how is done now, connection all the parts one after another, maybe just connect the `srv` and then all the rest of the things use the `app.Srv().Whatever`.

File: app/server.go (hunk 2 above)
Comment (jespino) [reply app/server.go:59]: Actually is part of the server, if we want to move it out of the app layer, the expectation is this is public, not private.

File: app/enterprise_test.go (hunk 4 above)
Comment (jespino) [reply app/enterprise_test.go:121]: I prefer to keep the separated, at least for now. The server is initialized a the launch of the application and the app, at the first app instantiation, so are two different moments.

Comment (agnivade) [reply app/enterprise_test.go:121]: Ok

File: app/options.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (agnivade) [reply app/options.go:105]: > maybe just connect the srv and then all the rest of the things use the app.Srv().Whatever.

Yes exactly. That's what I mean. All server acceses should be through the server. There is no need to bring up these fields once again at an app level. It breaks the abstraction and is confusing. Let me know if you want me to file a ticket for this.

File: app/app.go (hunk 3 above)
Comment (agnivade) [reply app/app.go:131]: Filed: https://mattermost.atlassian.net/browse/MM-25744

**Comment by jespino** [comment]
@lindy65 this PR is a very fundamental change, I think we should make some smoke testing around the correct behavior of the application, including searches, reads, writes, probably something that needs jobs, like compliance exports... After that, I would merge it on master and let is soak.

**Comment by agnivade** [comment]
Great stuff. Very excited !

**Comment by lindy65** [comment]
Thanks @jespino - will test as you've suggested.  

Please link me to the Jira ticket for this PR?  Thanks!

**Comment by jespino** [comment]
@lindy65 actually, there is no jira ticket for that, that is a continuation of this ticket https://mattermost.atlassian.net/browse/MM-23561

I can create one for you if you want.

**Comment by jespino** [comment]
Ticket create: https://mattermost.atlassian.net/browse/MM-25908

**Comment by lindy65** [comment]
Thanks @jespino 👍 

When you say "account migrations" (on the Jira ticket) what exactly do you mean?

**Comment by jespino** [comment]
@lindy65 good question, I mean the `account_migrations` code in the `enterprise` repository, that doesn't mean necessarily anything to you, so that is the code responsible to migrate from LDAP to SAML and from SAML to LDAP accounts, normally using the `mattermost` cli command (which is not `mmctl`).

**Comment by lindy65** [comment]
Hi @jespino 

**1. I've found a weird issue where not all messages sent from `sysadmin` are visible to `user-1`**

This is what `sysadmin` posted:
image

This is what `user-1` sees:
image

If I refresh `user-1's` screen, then all the posts appear:
image

Since I refreshed `user-1's` screen I am not able to reproduce this again. If it's helpful, this is what is in the js.console:
image

**2. Another weird issue is ` sysadmin`  posts the numbers 1-10 in succession - so 10 different posts one after the other - and in some channels it posts one after the other but in other channels it posts like this:**
image

Here is a similar post where it posts one after the other:
image

**3. When `sysadmin`  updates the header in a channel, she sees the system message that the header has been updated but `user-1`  doesn't see the system message:** **_This is fixed by refreshing `user-1's` screen._**
Sysadmin's view:
image

User-1's view:
image

**4. Also with system messages, when `sysadmin`  added ` user-1`  to a channel neither ` sysadmin` nor `user-1`  see a system post:** **_This is fixed by refreshing the screens for both users_**

Sysadmin's view:
image

**5. The same is true for removing a user from a channel. When `sysadmin` removed `user-1` from the channel, channel members icon still shows 2 members and the system post is not posted in the channel:** **_This is fixed by refreshing the screen_**

Sysadmin's view:
image

Sysadmin's view after refresh:
image

**6. The X to clear search results in the System Console is broken.  If I click on the weird characters, the search is still cleared though:** **_NOTE: Sometime during testing this fixed itself and the X is now visible and correct??_**

image

**7. After setting Announcement Banner to True, typing in some text and Saving, the banner does not appear for the Sysadmin:** **_This is NOT fixed by refreshing the screen.  It did fix itself once I updated another setting in the System Console and saved - then Sysadmin sees the Announcement Banner_**

Sysadmin's view:
image

**The banner DOES appear for User-1 though**.  

I'm not 100% sure but **could it be a bug with another banner being visible to the Sysadmin before setting the Announcement Banner to True**?  

In this case, there is an "expiring licence" banner on the spinwick that only the Sysadmin sees.  After dismissing that banner, the Sysadmin does not see the Announcement Banner.  I tried both scenarios from Sysadmin account - dismissing licence expiry banner first and then setting up Announcement Banner as well as **not** dismissing licence expiry banner and setting up Announcement Banner but neither scenario is any different and Sysadmin still does NOT see the Announcement Banner.

User-1's view:
image

**8. After updating Announcement Banner to be able to dismiss it and editing the text of the banner, Sysadmin sees the changes but User-1 does not:**  **_This is NOT fixed by User-1 refreshing the screen.  It IS fixed for User-1 by logging out and back in._**

Sysadmin's view:
image

User-1's view:
image

**9. Should the server logs be totally blank?:**

image

**10. I set off a CSV Compliance Export but it stayed in the "Pending" state for over half an hour:**
image

**Comment by jespino** [comment]
Can we try to verify this bugs in other branch to discard is related to my changes? Some of them sounds related to my changes, some of them not. I want to confirm if that is something happening in master, or is because my changes.

**Comment by lindy65** [comment]
Hi @jespino - I am NOT able to reproduce any of the above issues on `master.test`

**Comment by jespino** [comment]
Ok, I'll review it, thanks!

**Comment by jespino** [comment]
@lindy65 I will check it locally, I think i can have to do with the cloud testing environment, but I have to verify what is because of that, and what is because of my changes (maybe all are because of my changes).

**Comment by lindy65** [comment]
Thanks @jespino - is this ready for retesting then?

**Comment by jespino** [comment]
Yes, I think so, I'm dealing with some minor problems on the tests, but should be fixed tomorrow.

**Comment by lindy65** [comment]
@jespino - all issues are fixed but I was not able to test 7, 8 and 10 on the spinwick as it's a team edition one.  I will try again to spin up an E20 spinwick.

**Comment by jespino** [comment]
@lindy65 probably I'll need to fix it the we build before you can test it. Let me check it tomorrow. Today is bank holiday here

**Comment by jespino** [comment]
@lindy65 this should be ready now

**Comment by jespino** [comment]
Ohh yes, thanks @lindy65 really looking forward to get this merge into master :)

**Pull request merged**